plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringPlugin.java

```diff
@@ -57,34 +65,64 @@
 
     public void init() {
         LOGGER.info("Spring plugin initialized");
+        this.registerResourceListeners();
+        this.registerBasePackageFromConfiguration();
     }
     public void init(String version) {
         LOGGER.info("Spring plugin initialized - Spring core version '{}'", version);
+        this.registerResourceListeners();
+        this.registerBasePackageFromConfiguration();
+    }
+
+    private void registerResourceListeners() {
+        WatchResourcesClassLoader.getDefault(appClassLoader).addWatchResourcesListener(new WatchResourcesListener() {
+            @Override
+            public void onFileChange(URL url) {
+                scheduler.scheduleCommand(new XmlBeanRefreshCommand(appClassLoader, url));
+            }
+        });
     }
 
     /**
-     * Register both hotswap transformer AND watcher - in case of new file the file is not known
-     * to JVM and hence no hotswap is called. The file may even exist, but until is loaded by Spring
-     * it will not be known by the JVM. File events are processed only if the class is not known to the
-     * classloader yet.
-     *
-     * @param basePackage only files in a basePackage
+     * register base package prefix from configuration file
      */
-    public void registerComponentScanBasePackage(final String basePackage) {
-        LOGGER.info("Registering basePackage {}", basePackage);
+    public void registerBasePackageFromConfiguration() {
+        if (basePackagePrefixes != null) {
+            for (String basePackagePrefix : basePackagePrefixes) {
+                this.registerBasePackage(basePackagePrefix);
+            }
+        }
+    }
+
+    private void registerBasePackage(final String basePackage) {
         final SpringChangesAnalyzer analyzer = new SpringChangesAnalyzer(appClassLoader);
         hotswapTransformer.registerTransformer(appClassLoader, basePackage + ".*", new ClassFileTransformer() {
             @Override
             public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
                 if (classBeingRedefined != null) {
+                    // any annotation with array will lead to xxx
```

Comment:
Thanks for your advice:
* I don't know @OnResourceFileEvent before, I will change into using that (Still, do you think I should preserve default instance of WatchResourcesListener?).
* basePackagePrefix is used when component-scan is turned off and even if it's turned on we can use it  to filter bean we should create proxy for to improve performance (filtering third party lib beans), I will elaborate the doc.
* At first I designed to filter xml file using XmlBeanDefinationScannerAgent.instances, then I realize that url can be different if we edit xml from IDE while running it on server, now I plan to use classpath url as key so they are same (Still haven't got a better means to get classpath url for changed file other than get substring after "classes/" or "src/main/resources")

I will work on these in next a fews days.

Reply:
* I would suggest to remove WatchResourcesListener, because it is duplicate to current functionality (if you do not have any use case that cannot be covered by HA annotations)
* basePackagePrefix - Ok
* I think also that it is not possible to match the Spring configuration file to XmlBeanDefinationScannerAgent.instances. Probably the correct solution would be to immitate Spring configuration file discovery (hook into ). If there is a small performance impact, your solution may be appropriate, but we should check it for multiple resource file redefinition.

Please replace the WatchResourcesListener and I will merge your changes. Further enhancements can be solved in subsequent commits. Thank you.